snippets/mongocompat/mongotypes.js

```diff
@@ -1,5 +1,60 @@
 // Date and time types
 if (typeof (Timestamp) != "undefined") {
+    const OriginalTimestamp = Timestamp;
+
+    // Reference: https://github.com/10gen/mongo/blob/master/src/mongo/scripting/mozjs/timestamp.cpp:67-78
+    function validateTimestampComponent(component, name) {
+        const MAX_UINT32 = 4294967295; 
+
+        if (typeof component !== 'number') {
+            throw new Error(name + " must be a number");
+        }
+
+        const val = Math.floor(component);
+        if (val < 0 || val > MAX_UINT32) {
+            throw new Error(
+                name + " must be non-negative and not greater than " + MAX_UINT32 + ", got " + val
+            );
+        }
+
+        return val;
+    }
+
+    Timestamp = function(t, i) {
+        if (arguments.length === 0) {
+            return new OriginalTimestamp({ t: 0, i: 0 });
+        }
+
+        if (arguments.length === 1) {
+            const proto = Object.getPrototypeOf(t);
+            if ((proto === null || proto === Object.prototype) && ('t' in t || 'i' in t)) {
+                var validatedT = validateTimestampComponent(t.t || 0, "Timestamp time (seconds)");
+                var validatedI = validateTimestampComponent(t.i || 0, "Timestamp increment");
+                return new OriginalTimestamp({ t: validatedT, i: validatedI });
+            }
+            return new OriginalTimestamp(t);
+        }
+
+        // Reference: https://github.com/10gen/mongo/blob/master/src/mongo/scripting/mozjs/timestamp.cpp:91-98
```

Comment:
```suggestion
        // Reference: https://github.com/mongodb/mongo/blob/c4d21d3346572e28df2f174df4d87e7618df4a77/src/mongo/scripting/mozjs/timestamp.cpp#L91-L98
```

- GH supports built-in line references via `#` anchor tags
- We should use permalinks instead of references to branches, because the exact lines (or even the filenames) will otherwise inevitably change over time
- We should always link to the public repository if we can, since this is also a public repository